User Listing application

Sandesh3003 opened the pull request:
Assigned task: Create a form to create user using ftl and then list the users by fetching data from groovy and also apply the filters while listing users.

Sandesh3003 committed:
Created a view entity UserAndParty for fetching user data

Created a view UserAndParty to fetch data from multiple entities UserLogin, Person, ContactMech, PartyContactMech, PostalAddress, TelecomNumber, PartyContactMechPurpose, UserLoginSecurityGroup

Sandesh3003 committed:
Created a view entity UserAndParty for fetching user data

Created a view UserAndParty to fetch data from multiple entities UserLogin, Person, ContactMech, PartyContactMech, PostalAddress, TelecomNumber, PartyContactMechPurpose, UserLoginSecurityGroup

Sandesh3003 committed:
Created a create user form and list screen with custom decorator and related events and services

Created a CreateUser.ftl to create and ListUser.ftl for listing users and created event createUserEvent.java in UserEvents.java to validate the create form inputs and call the wrapper service createUserByJavaService which calls services createPersonAndUserLogin to create user and person, createPostalAddress to create(insert) the address(s) of the user, createTelecomNumber service to create(insert) the contact number(s) of the user and createEmailAddress to create(insert) email of the user. The definition of the createUserByJavaService can be found in services_view.xml

Sandesh3003 committed:
Created a create user form and list screen with curtom decorator and related events and services

Created a CreateUser.ftl to create and ListUser.ftl for listing users and created event createUserEvent.java in UserEvents.java to validate the create form inputs and call the wrapper service createUserByJavaService which calls services createPersonAndUserLogin to create user and person, createPostalAddress to create(insert) the address(s) of the user, createTelecomNumber service to create(insert) the contact number(s) of the user and createEmailAddress to create(insert) email of the user. The definition of the createUserByJavaService can be found in services_view.xml

Sandesh3003 committed:
Created a create user form and list screen with custom decorator and related events and services

Created a CreateUser.ftl to create and ListUser.ftl for listing users and created event createUserEvent.java in UserEvents.java to validate the create form inputs and call the wrapper service createUserByJavaService which calls services createPersonAndUserLogin to create user and person, createPostalAddress to create(insert) the address(s) of the user, createTelecomNumber service to create(insert) the contact number(s) of the user and createEmailAddress to create(insert) email of the user. The definition of the createUserByJavaService can be found in services_view.xml

Sandesh3003 committed:
Merge branch 'usermgr' of https://github.com/Sandesh3003/ofbiz-tutorial into usermgr

Sandesh3003 committed:
Created services definition for the Java service implementation

Write the definition for the Java services createPostalAddress (def: createPostalAddressJava), createTelecomNumber (def: createTelecomNumberJava) and createEmailAddress (def: createEmailAddressJava)

Sandesh3003 committed:
Improved UI by making create form and list form responsive and separate screens for the create user and list users

abhishek-verma-hotwax commented on entitydef/user-entitymodel_view.xml:
```diff
@@ -0,0 +1,61 @@
+        <member-entity entity-alias="CM" entity-name="ContactMech"/>
+        <member-entity entity-alias="PCMP" entity-name="PartyContactMechPurpose"/>
+        <member-entity entity-alias="ULSG" entity-name="UserLoginSecurityGroup"/>
+        <alias name="userLoginId" entity-alias="UL" field="userLoginId" prim-key=""/>
```
Why we use prim-key in view entity ?

Sandesh3003 replied:
The fields that formed the primary key of the member entities are included in the primary key of the view-entity. The prim-key attribute is used to identify such fields(alias) explicitly. If not specified, the Entity Engine will examine the field's definition from which it originated(member-entity) to determine if it is part of the original entity's(member-entity's) primary key.

**Note:** There were discrepancies in the earlier commit of the view-entity "UserAndParty". In the recent commit "Created User Profile screen" those discrepancies have been eliminated.

abhishek-verma-hotwax replied:
we don't have to add this prim-key tag to view entity

also when you add alias in the view entity do like this 
<alias entity-alias="UL" field="userLoginId" name="userLoginId"/>
so if the field and name is same then we do not need to add the "filed"  tag.

Sandesh3003 replied:
Yes sir I have removed the field attribute from the alias tag wherever I am using the same alias name as of field.

abhishek-verma-hotwax commented on entitydef/user-entitymodel_view.xml:
```diff
@@ -0,0 +1,61 @@
+        <alias name="areaCode" entity-alias="TN"/>
+        <alias name="contactNumber" entity-alias="TN"/>
+        <alias name="groupId" entity-alias="ULSG"/>
+        <view-link entity-alias="UL" rel-entity-alias="P" rel-optional="true">
```
why we set rel-optional="true" here?

abhishek-verma-hotwax replied:
same question follows below

Sandesh3003 replied:
The "rel-optional" attribute is used to indicate whether the related entity is optional or not. In this case, "rel-optional='true'" is used because it is possible that certain "userLogin" may not be associated with a "party" but still exists. Therefore, in such situations, we still want to retrieve the records. The "rel-optional='true'" performs a left outer join and we get the records for the "userLogin" details with NULL in the "party" details. Similarly, not all contact mechanisms may be associated with all postal addresses, telecom numbers, or contact mech purposes. That's why we use "rel-optional='true'".

abhishek-verma-hotwax commented on servicedef/services_user.xml:
```diff
@@ -0,0 +1,59 @@
+    <version>1.0</version>
+    <service name="fetchAllUserData" engine="java" location="org.apache.ofbiz.ofbiztutorial.usermgr.services.UserMgrServices"
+             invoke="fetchAllUserData" auth="true" default-entity-name="UserAndParty">
+        <auto-attributes mode="IN" include="nonpk" optional="true"/>
```
Why we didn't include PK here ?

Sandesh3003 replied:
This section of the code is unnecessary. I previously wrote a service to retrieve data, but I now use a groovy script to fetch it directly. This section will be removed in the next commit.

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/events/UserEvents.java:
```diff
@@ -0,0 +1,132 @@
+        // Extract data from the form
+        HttpSession session = request.getSession();
+        GenericValue userLogin = (GenericValue) session.getAttribute("userLogin");
+        System.out.println("userLogin "+userLogin+"========================");
```
remove system logs from here

Sandesh3003 replied:
Will be removed in the next commit.

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/events/UserEvents.java:
```diff
@@ -0,0 +1,132 @@
+        String lastName = request.getParameter("lastName");
+        String middleName = request.getParameter("middleName");
+//        String gender = request.getParameter("gender");
+<<<<<<< HEAD
```
you've pushed the conflicts please remove this.

Sandesh3003 replied:
Removed these conflicts in the recent commit.

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/events/UserEvents.java:
```diff
@@ -0,0 +1,132 @@
+//        if(UtilValidate.isEmpty(gender)){
+//            return ServiceUtil.returnError("Gender cannot be empty!!").toString();
+//        }
+<<<<<<< HEAD
```
you've pushed the conflicts please resolve this.

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/events/UserEvents.java:
```diff
@@ -0,0 +1,132 @@
+//        }
+=======
+        if (!UtilValidate.isDate(birthDate)) {
+            return ServiceUtil.returnError("Incorrect Date Format in Birth Date!!!" + birthDate).toString();
```
Add all errors in a list and return it only one time with all the issues

Sandesh3003 replied:
I'll cover the same in the next commit.

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/events/UserEvents.java:
```diff
@@ -0,0 +1,132 @@
+            serviceParams.put("firstName", firstName);
+            serviceParams.put("lastName", lastName);
+            serviceParams.put("middleName", middleName);
+<<<<<<< HEAD
```
Please resolve this conflics

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/services/UserMgrServices.java:
```diff
@@ -0,0 +1,164 @@
+import org.apache.ofbiz.service.DispatchContext;
+import org.apache.ofbiz.service.GenericServiceException;
+import org.apache.ofbiz.service.ServiceUtil;
+<<<<<<< HEAD
```
resolve conflicts

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/services/UserMgrServices.java:
```diff
@@ -0,0 +1,164 @@
+        String emailAddress = (String) context.get("emailAddress");
+        GenericValue userLogin = (GenericValue) context.get("userLogin");
+
+        System.out.println("============="+userLogin+"===============");
```
Please remove Logs

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/services/UserMgrServices.java:
```diff
@@ -0,0 +1,164 @@
+        }
+
+        // Invoke createPersonAndUserLogin service to create user and associated person
+        Map<String, Object> createPersonParams = Map.of(
```
use Ofbiz method instead of Map.of

UtilMisc.toMap

Sandesh3003 replied:
Sure sir I'll implement the change in the next commit.

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/services/UserMgrServices.java:
```diff
@@ -0,0 +1,164 @@
+            return createPersonResult;
+        }
+        GenericValue newUserLogin = (GenericValue) createPersonResult.get("newUserLogin");
+<<<<<<< HEAD
```
resolve conflicts

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/services/UserMgrServices.java:
```diff
@@ -0,0 +1,164 @@
+        }
+        GenericValue newUserLogin = (GenericValue) createPersonResult.get("newUserLogin");
+<<<<<<< HEAD
+        Map<String, Object> addUserLoginSecurityGroup = Map.of(
```
use Ofbiz method instead of Map.of

UtilMisc.toMap

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/services/UserMgrServices.java:
```diff
@@ -0,0 +1,164 @@
+=======
+>>>>>>> b5bdf4e25b3ebee19ea8ad8ad27302fa32a4f9f6
+        // Invoke createPostalAddress service
+        Map<String, Object> createPostalAddressParams = Map.of(
```
use Ofbiz method instead of Map.of

UtilMisc.toMap

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/services/UserMgrServices.java:
```diff
@@ -0,0 +1,164 @@
+                "city", city,
+                "postalCode", postalCode,
+                "countryGeoId", countryGeoId,
+<<<<<<< HEAD
```
resolve conflicts

abhishek-verma-hotwax commented on webapp/ofbiz-tutorial/WEB-INF/controller.xml:
```diff
@@ -53,7 +53,24 @@ under the License.
+    <request-map uri="createUserEvent">
+        <security https="true" auth="true"/>
+        <event type="java" path="org.apache.ofbiz.ofbiztutorial.usermgr.events.UserEvents" invoke="createUserEvent"/>
+        <response name="success" type="view" value="CreateUserFtl" save-home-view="true"/>
```
what is save-home-view="true" here

abhishek-verma-hotwax replied:
can you please elaborate

Sandesh3003 replied:
By using the save-view-home attribute, you can save your current request view for later access. This feature is handy when you need to retrieve the view-home, representing the last saved home position, in a response. Opting for save-last-view="true" would be more fitting here, as it doesn't refer to the home (main page) position in this context.

abhishek-verma-hotwax commented on webapp/ofbiz-tutorial/WEB-INF/web.xml:
```diff
@@ -52,7 +52,7 @@ under the License.
         <init-param>
             <param-name>allowedPaths</param-name>
-            <param-value>/error:/control:/select:/index.html:/index.jsp:/default.html:/default.jsp:/images</param-value>
+            <param-value>/error:/control:/select:/index.html:/index.jsp:/default.html:/default.jsp:/images:/usermgr/css:/usermgr/js:/usermgr/includes</param-value>
```
do we really need to change in to web.xml

Sandesh3003 replied:
To include the external CSS and JS we were required to make changes in the web.xml to allow access to those resources.

abhishek-verma-hotwax commented on widget/CommonScreens.xml:
```diff
@@ -40,28 +40,41 @@ under the License.
         </section>
     </screen>
-
     <screen name="Ofbiz-tutorialCommonDecorator">
```
do we really need to change the Ofbiz-tutorialCommonDecorator decorator

Sandesh3003 replied:
Ofbiz-tutorialCommonDecorator screen serves as a custom decorator for my application. We could also create another screen to serve as a custom decorator. Hence, it is not necessarily required to change the Ofbiz-tutorialCommonDecorator.

abhishek-verma-hotwax commented on servicedef/services_user.xml:
```diff
@@ -0,0 +1,59 @@
+    </service>
+    <service name="createUserByJavaService" engine="java" location="org.apache.ofbiz.ofbiztutorial.usermgr.services.UserMgrServices"
+             invoke="createUserByJavaService" auth="false" export="true" action="POST">
+        <attribute name="userLogin" type="Map" mode="IN"/>
```
Please use implementes tag instead of adding attributes seprately

Sandesh3003 replied:
Sure I'll include in the next commit.

Sandesh3003 committed:
Created User Profile screen

Sandesh3003 committed:
Removed unneccessary logs

Sandesh3003 committed:
Removed un-necessary code

Sandesh3003 committed:
Replaced Map.of() method with UtilMisc.toMap() method of ofbiz

Sandesh3003 committed:
Reduced code redundancy in service definition by implementing predefined services and preparing context using ofbiz method makeValidContext

abhishek-verma-hotwax commented on servicedef/services_user.xml:
```diff
@@ -0,0 +1,61 @@
+        <implements service="createPostalAddress"/>
+        <implements service="createTelecomNumber"/>
+        <attribute name="emailAddress" type="String" mode="IN" optional="false"/>
+<!--        <attribute name="userLogin" type="Map" mode="IN"/>-->
```
please remove commets

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/events/UserEvents.java:
```diff
@@ -0,0 +1,130 @@
+        HttpSession session = request.getSession();
+        GenericValue userLogin = (GenericValue) session.getAttribute("userLogin");
+        String userLoginId = request.getParameter("userLoginId");
+        String currentPassword = request.getParameter("password");
```
Explore this method and use it here 

UtilHttp.getCombinedMap()

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/events/UserEvents.java:
```diff
@@ -0,0 +1,130 @@
+            serviceParams.put("userLogin", userLogin);
+            serviceParams.put("userLoginId", userLoginId);
+            serviceParams.put("currentPassword", currentPassword);
+            serviceParams.put("currentPasswordVerify", currentPasswordVerify);
```
please use makeValidContext method here

Sandesh3003 replied:
To use the makeValidContext method first I have to create a context map and put all these values in the context map. Then pass that context in the makeValidContext method to get valid context for the service.

abhishek-verma-hotwax commented on src/main/java/org/apache/ofbiz/ofbiztutorial/usermgr/services/UserMgrServices.java:
```diff
@@ -0,0 +1,138 @@
+        }
+
+        GenericValue newUserLogin = (GenericValue) createPersonResult.get("newUserLogin");
+        Map<String, Object> addUserLoginSecurityGroup = UtilMisc.toMap(
```
Please use UtilMisc method to creating map and list

Sandesh3003 committed:
Updated UserEvents.java to use UtilHttp.getCombinedMap() to create context

Sandesh3003 committed:
Updated: UserMgrServices.java to enlose the wrapper services to create user and add user detals in a single try-catch block

Sandesh3003 committed:
Improved: update in service definition to accept partyId in input parameters and update in service implementation createUserByJavaService

Updated service definition of the createPostalAddressJava, createTelecomNumberJava and createEmailAddressJava to accept partyId as input parameter.

Sandesh3003 committed:
Fixed: UI fixes to display data not found content if user list is empty and include gender field in the create form
